### PR TITLE
Major update for Android runtime support and camera devices.

### PR DESCRIPTION
**Pull request by koush**
# Pull Request Checklist

Please confirm that you've done the following when opening a new pull request:

  - [X ] For fixes and other improvements, please reference the GitHub issue that your change addresses.
  - [X ] For fixes, optimizations and new features, please add an entry to the CHANGES.md file.
  - [X ] Run mvn compile before committing, so that the auto-code formatter will format your changes consistently with the rest of the project.

### CODE REVIEW

**Comment by koush** [comment]
I had forked this repository a couple years ago and never contributed back. I basically hand picked all my changes/fixes and cleaned them up to apply them to this repo (which had a large refactor and package rename).

I'll leave comments inline.

**Comment by koush** [comment]
Unclear why travis is failing. I think the gradle wrapper is correct and working properly on a clean checkout from my machine.

**Comment by koush** [comment]
Fixes this. https://github.com/hap-java/HAP-Java/issues/117

**Comment by yfre** [comment]
@koush thank you for PRs. 
it has many very good things, e.g.
- upgrade of bouncycastle which we have tried several times without success (see #80)
- camere support
- async http request handling
- android support 

maybe you could it split in several PRs so that it is easier to review and easier to get CI working

**Comment by koush** [comment]
I doubt I'll have the time to break this up into separate changes, as I had to basically implement it as one big patch to simply get everything working again (due to the package rename). The old cherry picks are for the pre-rename code.

I think the reason CI is failing is because it detects the build.gradle which is necessary for compiling+debugging on android. I think deleting that file would fix the CI. The CI needs to ignore it. It may be possible to delete it, but then Android would not compile. the mvn built jars should work on Android too though.

The patch is fairly non-intrusive and actually not that large. The majority of the files are the gradle wrapper and the various data types for video (which are unused anyways).

**Comment by gjvanderheiden** [comment]
I agree this could use a split up. 

BouncyCastle could be updated to 1.67, with a small modification
if (!Arrays.constantTimeAreEqual(calculatedMAC, receivedMAC)) {
      throw new TlsFatalAlert(AlertDescription.bad_record_mac);
  }
TlsFatalAlert could be converted into IOException, because TlsFatalAlert doesn't exist anymore in .67. TlsFatalAlert is just an IOException, no biggie.

I also was a bit surprised there is no TLV object in HAP-java. You introduced it, but now it isn't used in the server packages.
So it is a really nice addition and should be exploited in the server package with some extra methods (getState(), getMethod() etc).

**Comment by gjvanderheiden** [comment]
Could you maybe rename the gradle file for now, so that the build doesn't fail?

File: src/main/java/io/github/hapjava/characteristics/impl/base/BaseCharacteristic.java [hunk 1]
```diff
@@ -113,7 +113,7 @@ public BaseCharacteristic(
                       .add("type", shortType)
                       .add("perms", perms.build())
                       .add("format", format)
-                      .add("ev", false)
+                      //                      .add("ev", false)
```
Comment (gjvanderheiden) [inline src/main/java/io/github/hapjava/characteristics/impl/base/BaseCharacteristic.java:116]: Was it your intention to comment  this out? Looks like a rather common piece of code.

File: pom.xml [hunk 2]
```diff
@@ -101,19 +101,19 @@
 		<dependency>
 			<groupId>com.nimbusds</groupId>
 			<artifactId>srp6a</artifactId>
-			<version>1.5.2</version>
+			<version>2.0.2</version>
```
Comment (gjvanderheiden) [inline pom.xml:104]: 2.1.0 is the newest

File: pom.xml [hunk 3]
```diff
@@ -101,19 +101,19 @@
...
 		</dependency>
 
 		<dependency>
 			<groupId>org.bouncycastle</groupId>
 			<artifactId>bcprov-jdk15on</artifactId>
-			<version>1.51</version>
+			<version>1.60</version>
```
Comment (gjvanderheiden) [inline pom.xml:110]: 1.67 is the newest

File: README.md [hunk 4]
```diff
@@ -32,7 +32,7 @@ Current implementation fully supports 37 HomeKit accessory/services.
 |  Air Quality Sensor               |     :white_check_mark:    |
 |  Audio Stream Management          |     :x:                   |
 |  Battery Service                  |     :white_check_mark:    |
-|  Camera RTP Stream Management     |     :x:                   |
+|  Camera RTP Stream Management     |      :white_check_mark:   |
```
Comment (gjvanderheiden) [inline README.md:35]: needs update from hap-java/hap-java 

And this text needs update:
Current implementation fully supports 37 HomeKit accessory/services.

File: CHANGES.md [hunk 5]
```diff
@@ -27,3 +27,5 @@
 * Occupancy sensor support [#59](https://github.com/hap-java/HAP-Java/pull/59)
 * Leak sensors and valve support added [#52](https://github.com/hap-java/HAP-Java/pull/52)
 * Notifications are batched now, when possible [#66](https://github.com/hap-java/HAP-Java/pull/66)
+# Android Support
+# Camera Support
```
Comment (gjvanderheiden) [inline CHANGES.md:31]: should be an asterisk instead of a hash

File: CHANGES.md [hunk 6]
```diff
@@ -27,3 +27,5 @@
 * Occupancy sensor support [#59](https://github.com/hap-java/HAP-Java/pull/59)
 * Leak sensors and valve support added [#52](https://github.com/hap-java/HAP-Java/pull/52)
 * Notifications are batched now, when possible [#66](https://github.com/hap-java/HAP-Java/pull/66)
+# Android Support
```
Comment (gjvanderheiden) [inline CHANGES.md:30]: should be an asterisk instead of a hash

File: src/main/java/io/github/hapjava/server/impl/crypto/ChachaEncoder.java [hunk 7]
```diff
@@ -39,9 +39,8 @@ private KeyParameter initRecordMAC(ChaChaEngine cipher) {
     byte[] firstBlock = new byte[64];
     cipher.processBytes(firstBlock, 0, firstBlock.length, firstBlock, 0);
 
-    // NOTE: The BC implementation puts 'r' after 'k'
-    System.arraycopy(firstBlock, 0, firstBlock, 32, 16);
-    KeyParameter macKey = new KeyParameter(firstBlock, 16, 32);
+    // NOTE: The older BC implementation puts 'r' after 'k'0
```
Comment (gjvanderheiden) [inline src/main/java/io/github/hapjava/server/impl/crypto/ChachaEncoder.java:42]: Is this comment still applicable?

File: src/main/java/io/github/hapjava/server/impl/crypto/ChachaDecoder.java [hunk 8]
```diff
@@ -45,9 +45,8 @@ private KeyParameter initRecordMAC(ChaChaEngine cipher) {
     byte[] firstBlock = new byte[64];
     cipher.processBytes(firstBlock, 0, firstBlock.length, firstBlock, 0);
 
-    // NOTE: The BC implementation puts 'r' after 'k'
-    System.arraycopy(firstBlock, 0, firstBlock, 32, 16);
-    KeyParameter macKey = new KeyParameter(firstBlock, 16, 32);
+    // NOTE: The older BC implementation puts 'r' after 'k'
```
Comment (gjvanderheiden) [inline src/main/java/io/github/hapjava/server/impl/crypto/ChachaDecoder.java:48]: is this comment still applicable?

File: src/main/java/io/github/hapjava/server/impl/json/AccessoryController.java [hunk 9]
```diff
@@ -58,8 +59,9 @@ public HttpResponse listing() throws Exception {
     }
 
     try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
-      Json.createWriter(baos)
-          .write(Json.createObjectBuilder().add("accessories", accessories).build());
+      JsonWriter writer = Json.createWriter(baos);
+      writer.write(Json.createObjectBuilder().add("accessories", accessories).build());
+      writer.close();
```
Comment (gjvanderheiden) [inline src/main/java/io/github/hapjava/server/impl/json/AccessoryController.java:64]: JsonWriter is an AutoClosable, makes more sense to use the try block:

try {(ByteArrayOutputStream baos = new ByteArrayOutputStream();
        JsonWriter writer = Json.createWriter(baos))

File: src/main/java/io/github/hapjava/server/impl/json/CharacteristicsController.java [hunk 10]
```diff
@@ -59,9 +59,10 @@ public HttpResponse get(HttpRequest request) throws Exception {
       }
     }
     try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
-      Json.createWriter(baos)
-          .write(
-              Json.createObjectBuilder().add("characteristics", characteristics.build()).build());
+      JsonWriter writer = Json.createWriter(baos);
+      writer.write(
+          Json.createObjectBuilder().add("characteristics", characteristics.build()).build());
+      writer.close();
```
Comment (gjvanderheiden) [inline src/main/java/io/github/hapjava/server/impl/json/CharacteristicsController.java:65]: JsonWriter is an AutoClosable, makes more sense to use the try block:

try {(ByteArrayOutputStream baos = new ByteArrayOutputStream();
        JsonWriter writer = Json.createWriter(baos))

**Pull request closed without merging**